Sources/SwiftParser/Lexer/Cursor.swift

```diff
@@ -133,11 +133,61 @@ extension Lexer.Cursor {
       case .afterStringLiteral: return false
       case .afterClosingStringQuote: return false
       case .inStringInterpolationStart: return false
-      case .inStringInterpolation(stringLiteralKind: let stringLiteralKind, parenCount: _, stringInterpolationStart: _): return stringLiteralKind != .multiLine
+      case .inStringInterpolation(stringLiteralKind: let stringLiteralKind, parenCount: _): return stringLiteralKind != .multiLine
       case .afterStringInterpolation: return false
       }
     }
   }
+
+  /// A data structure that holds the state stack entries in the lexer. It is
+  /// optimized for situations where there's at most one state on the stack
+  /// (in addition to the bottom `normal` mode) and to not create any ARC
+  /// traffic.
+  ///
+  /// It does this by representing the bottom `normal` state implicitly and
+  /// storing one additional state inline. If the stack needs to contain more
+  /// entries, the required memory is allocated in a bump allocator that is
+  /// expected to outlive the stack.
+  struct StateStack {
+    private var topState: State? = nil
+    private var stateStack: UnsafeBufferPointer<State>? = nil
+
+    var currentState: State {
+      return topState ?? .normal
+    }
+
+    mutating func perform(stateTransition: Lexer.StateTransition, stateAllocator: BumpPtrAllocator) {
+      switch stateTransition {
+      case .push(newState: let newState):
+        if let topState = topState {
+          if let stateStack = stateStack {
+            let newStateStack = stateAllocator.allocate(State.self, count: stateStack.count + 1)
+            let (_, existingStateStackEndIndex) = newStateStack.initialize(from: stateStack)
+            newStateStack[existingStateStackEndIndex] = topState
+            self.stateStack = UnsafeBufferPointer(newStateStack)
+          } else {
+            let newStateStack = stateAllocator.allocate(State.self, count: 1)
+            newStateStack[0] = topState
+            self.stateStack = UnsafeBufferPointer(newStateStack)
+          }
+        }
+        topState = newState
+      case .replace(newState: let newState):
+        topState = newState
```

Comment:
_If_ we place the `topState` into `stateStack`, we need to replace the `stateStack.last` too.
otherwise `push` `replace` `push` `pop` operation doesn't restore the replaced state.

Reply:
When I implemented this, I measured and found that storing `topState` inline gives the best performance because that way we don’t need to allocate a new `stateStack` for every string literal (only for every interpolation, which I believe should be rarer). I also tried storing the top two states inline (so that we only need to allocate memory in the bump allocator for string literals inside interpolations) but that was slightly slower.

Reply:
Ditto. Please ignore this.